_posts/2016-04-19-MIR.md

```diff
@@ -0,0 +1,725 @@
+
+To handle this scenario today, the compiler uses *zeroing*. Or, more
+accurately, *overwriting*. What this means is that, if ownership of
+`data` is moved, we will overwrite the stack slot for `data` with a
+specific, distinctive bit pattern that is not a valid pointer (we used
+to use zeroes, so we usually call this zeroing, but we've since
+shifted to something different). Then, when it's time to free `data`,
+we check whether it was overwritten. (As an aside, this is actually
+roughly the same thing that the equivalent C++ code would do.)
+
+But we'd like to do better than that. What we would *like* to do is to
+using boolean flags on the stack that tell us what needs to be freed.
+So that might look something like this:
+
+```rust
+fn send_if(data: Vec<Data>) {
+    let mut data_is_owned = true;
+
+    if some_condition(&data) {
+        send_to_other_thread(data);
+        data_is_owned = false;
+    }
+
+    post_send();
+
+    // Free `data`, but only if we still own it:
+    if data_is_owned {
+        mem::drop(data);
+    }
+}
+```
+
+Of course, you couldn't write code like this in Rust. You're not
+allowed to acccess the variable `data` after the `if`, since it might
+have been moved. (This is yet another example of where we can do
+things in MIR that we would not want to allow in full Rust.)
+
+Using boolean stack flags like this has a lot of advantages. For one,
+it's more efficient: instead of overwriting the entire vector, we only
+have to set the one flag. But also, it's easier to optimize: imagine
+that, through inlining or some other means, the compiler was able to
+determine that `some_condition` would always be true. In that case,
+standard constant propagation techniques would tell us that
+`data_is_owned` is always false, and hence we can just optimize away
+the entire call to `mem::drop`, resulting in tighter code. See
+[RFC 320][rfc320] for more details on that.
+
+However, implementing this optimization properly on the current
+compiler architecture is quite difficult. With MIR, it becomes
+relatively straightforward. The MIR control-flow-graph tells us
+explicitly where values will be dropped and when. When MIR is first
+generated, we assume that dropping moved data has no effect -- roughly
+like the current overwriting semantics. So this means that the MIR for
+`send_if` might look like this (for simplicity, I'll ignore unwinding
+edges).
+
+<center>
+![Non-zeroing drop example][nzd]
+</center>
+
+<!--
+[ some_condition(&data)? ] -true-> [ send_to_other_thread(data) ]
+           |                              |
+         false                            |
+           |                              |
+           +----------+        +----------+
+                      |        |
+                      v        v
+                   [ post_send() ]
+                    [ drop(data) ]
+                      [ return ]
+-->
+
+We can then transform this graph by identifying each place where
```

Comment:
This paragraph and the next image don't appear in the rendered version, it looks like the previous comment isn't ended correctly: the closing `-->` appears to be converted into &ndash;> (`&ndash;>`).